Possible false positives using Objects.requireNonNull?

Hi there,

Consider the following code:

```
@Nullable
CustomValue value;    // member variable
...

public void doSomething() {
    value = dao.getValueWithId(1);    // getValue can return null if no object with the specified id exists
    Objects.requireNonNull(value, "Value is null!");    // throw NPE if value is null.  Afterwards, value can be considered non-null

    final String name = value.getName();
    ...
}
```

I see the following reports when running nullaway:

```
/Users/zach/Developer/Code/sampleapp/app/src/main/java/com/zao/sampleapp/SampleClass.java:20: warning: [NullAway] dereferenced expression value is @Nullable
        final String name = value.getName();
                                 ^
    (see http://t.uber.com/nullaway )
```

Is this a false positive? Should this be safely ignored?  Or am I using `Objects.requireNonNull` incorrectly?  I know, for instance, that `value = Objects.requireNonNull(dao.getValueWithId(1))` would avoid this, but I'd like to avoid "nesting" the calls.

Edit: I am using nullaway 0.6.4 - setup via errorprone in an Android project, FWIW.

Edit2: Updated `Objects.requireNonNull` call to include a message.  This works perfectly when the message parameter is omitted.

Are you on the latest release? I’m pretty sure this shouldn’t happen there

On Sat, Feb 2, 2019 at 8:51 PM Zach Olbrys <notifications@github.com> wrote:

> Hi there,
>
> Consider the following code:
>
> @Nullable
> CustomValue value;    // member variable
> ...
>
> public void doSomething() {
>     value = dao.getValueWithId(1);    // getValue can return null if no object with the specified id exists
>     Objects.requireNonNull(value);    // throw NPE if value is null.  Afterwards, value can be considered non-null
>
>     final String name = value.getName();
>     ...
> }
>
> I see the following reports when running nullaway:
>
> /Users/zach/Developer/Code/sampleapp/app/src/main/java/com/zao/sampleapp/SampleClass.java:20: warning: [NullAway] dereferenced expression value is @Nullable
>         final String name = value.getName();
>                                  ^
>     (see http://t.uber.com/nullaway )
>
> Is this a false positive? Should this be safely ignored? Or am I using
> Objects.requireNonNull incorrectly? I know, for instance, that value =
> Objects.requireNonNull(dao.getValueWithId(1)) would avoid this, but I'd
> like to avoid "nesting" the calls.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/uber/NullAway/issues/276>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALyUevRJ-6Flm9bvjch0R5ye25bRxLvks5vJmrIgaJpZM4af76_>
> .
>

--- a/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
+++ b/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
@@ -196,6 +196,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
         new ImmutableSetMultimap.Builder<MethodRef, Integer>()
             .put(methodRef("com.google.common.base.Preconditions", "<T>checkNotNull(T)"), 0)
             .put(methodRef("java.util.Objects", "<T>requireNonNull(T)"), 0)
+            .put(methodRef("java.util.Objects", "<T>requireNonNull(T,java.lang.String)"), 0)
             .put(methodRef("org.junit.Assert", "assertNotNull(java.lang.Object)"), 0)
             .put(
                 methodRef("org.junit.Assert", "assertNotNull(java.lang.String,java.lang.Object)"),
@@ -206,12 +207,12 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
                 methodRef(
                     "org.junit.jupiter.api.Assertions",
                     "assertNotNull(java.lang.Object,java.lang.String)"),
-                1)
+                0)
             .put(
                 methodRef(
                     "org.junit.jupiter.api.Assertions",
-                    "assertNotNull(java.lang.Object,java.util.function.Supplier<String>)"),
-                1)
+                    "assertNotNull(java.lang.Object,java.util.function.Supplier<java.lang.String>)"),
+                0)
             .build();
 
     private static final ImmutableSetMultimap<MethodRef, Integer> EXPLICITLY_NULLABLE_PARAMETERS =

--- a/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNativeModels.java
+++ b/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNativeModels.java
@@ -201,17 +201,22 @@ public class NullAwayNativeModels {
     }
   }
 
-  static void failIfNull(@Nullable Object o1, @Nullable Object o2) {
+  static void failIfNull(
+      @Nullable Object o1,
+      @Nullable Object o2,
+      @Nullable Object o3,
+      @Nullable Object o4,
+      @Nullable Object o5) {
     org.junit.Assert.assertNotNull(o1);
     o1.toString();
     org.junit.Assert.assertNotNull("Null!", o2);
     o2.toString();
-    org.junit.jupiter.api.Assertions.assertNotNull(o1);
-    o1.toString();
-    org.junit.jupiter.api.Assertions.assertNotNull(o2, "Null!");
-    o2.toString();
-    org.junit.jupiter.api.Assertions.assertNotNull(o2, () -> "Null!");
-    o2.toString();
+    org.junit.jupiter.api.Assertions.assertNotNull(o3);
+    o3.toString();
+    org.junit.jupiter.api.Assertions.assertNotNull(o4, "Null!");
+    o4.toString();
+    org.junit.jupiter.api.Assertions.assertNotNull(o5, () -> "Null!");
+    o5.toString();
   }
 
   static void nonNullParameters() {

--- a/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNegativeCases.java
+++ b/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNegativeCases.java
@@ -524,9 +524,11 @@ public class NullAwayNegativeCases {
     o.toString();
   }
 
-  static void requireNonNull(@Nullable Object o) {
+  static void requireNonNull(@Nullable Object o, @Nullable Object p) {
     Objects.requireNonNull(o);
     o.toString();
+    Objects.requireNonNull(p, "should be non null");
+    p.toString();
   }
 
   static void isEmpty(@Nullable String s) {

